blockchain/consensus/narwhal_cns/src/consensus.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+use async_trait::async_trait;
+use narwhal_fastcrypto::PublicKey;
+use std::fmt::Debug;
+use std::sync::Arc;
+use thiserror::Error;
+
+use forest_blocks::{Block, Tipset};
+use forest_chain::Error as ChainStoreError;
+use forest_chain::Scale;
+use forest_chain::Weight;
+use forest_chain_sync::consensus::Consensus;
+use forest_ipld_blockstore::BlockStore;
+use forest_state_manager::Error as StateManagerError;
+use forest_state_manager::StateManager;
+use fvm_ipld_encoding::Error as ForestEncodingError;
+use fvm_shared::bigint::BigInt;
+use nonempty::NonEmpty;
+
+#[derive(Debug, Error)]
+pub enum NarwhalConsensusError {
+    #[error("Chain store error: {0}")]
+    ChainStore(#[from] ChainStoreError),
+    #[error("StateManager error: {0}")]
+    StateManager(#[from] StateManagerError),
+    #[error("Encoding error: {0}")]
+    ForestEncoding(#[from] ForestEncodingError),
+    #[error("No miner address for validator key: {0}")]
+    NoMinerAddress(PublicKey),
+    #[error("{0}")]
+    Other(String),
+}
+
+/// In Narwhal Consensus we don't gossip blocks, we create deterministic
+/// blocks locally from the certificates we receive from Narwhal, and
+/// append them to our local chain.
+///
+/// The catch is that Narwhal only allows committee members to participate,
+/// so we basically cannot have non-validator nodes in the subnet.
+#[derive(Debug)]
+pub struct NarwhalConsensus;
+
+impl Scale for NarwhalConsensus {
+    fn weight<DB>(_: &DB, ts: &Tipset) -> anyhow::Result<Weight>
+    where
+        DB: BlockStore,
+    {
+        let header = ts.blocks().first().expect("Tipset is never empty.");
+        // We are building a single chain.
+        Ok(BigInt::from(header.epoch()))
+    }
+}
+
+#[async_trait]
+impl Consensus for NarwhalConsensus {
+    type Error = NarwhalConsensusError;
+
+    async fn validate_block<DB>(
+        &self,
+        _state_manager: Arc<StateManager<DB>>,
+        _block: Arc<Block>,
+    ) -> Result<(), NonEmpty<Self::Error>>
+    where
+        DB: BlockStore + Sync + Send + 'static,
+    {
+        // There is nothing to validate because we are only supposed to see blocks we created ourselves.
```

Comment:
I think option 1 would be the most straightforward one to have a fast verification, and allow blocks to sync from scratch, the problem I see is that we have a dynamic set of validators (they are allowed to join and leave anytime) so we have to be careful with how we validate that the set of signers for the checkpoint is still valid. 

Option 2 is probably the best, because IIRC from the Narwhal paper, blocks in a round include all the information needed to validate them in isolation, so even if Narwhal garbage collects blocks, as long as we are persisting the Narwhal blocks with their certificates locally in all peers to build the local chain, we should be able to request blocks from other peers (as we do know) from our current view to the heaviest block, and validate them (of course, this assumes that our stored blocks include all the information we need from Narwhal).

Something I have a hard time grasping every time I read the Narwhal paper is how one can be sure that all peers converge to the same local chain (I should probably read it again :) )

Reply:
You are right, option 1 isn't straight forward because you have to know the succession of committees for checkpoints to work. 

For option 2, currently I'm not saving the Narwhal committees, so we are talking about the Filecoin chains, hoping that a chain of unsigned blocks that many nodes agree on are trustworthy. You would not be able to validate the certificates because the Narwhal blocks are gone, and the invalid transactions from them are eliminated.

For the local ordering, note that what's in the Narwhal paper is Tusk, but what's implemented now is Bullshark, which says is's just 300 lines of embarrassingly simple pseudo code, but I haven't had the gumption to read those parts of the paper yet, it still looked complex.

Reply:
I see, so in option 2, even if we had information about Narwhal certificates, it'd be hard to validate the full chain.